### ISSUE TITLE
dart:ffi can't be imported when compiling to WASM. Flutter Web failed to compile to WASM.

### ISSUE THREAD

**Issue by adifyr** [flutter/flutter#149984]
### Steps to reproduce

0. I don't know if relevant or not - I have the `flutter_svg` dependency added for my web app.
1. Ran `flutter clean; flutter build web --wasm --release` in the terminal.

### Expected results

The Flutter Web App should compile properly with WebAssembly. At least it was working fine with Flutter Version `3.22.1`

### Actual results

Getting the following error:

```console
Target dart2wasm failed: ProcessException: Process exited abnormally with exit code 64:
*NOTE*: Compilation to WasmGC is experimental.
The support may change, or be removed, with no advance notice.

../../../../.pub-cache/hosted/pub.dev/vector_graphics_compiler-1.1.11+1/lib/src/svg/_tessellator_ffi.dart:6:1: Error: 'dart:ffi' can't be imported when
compiling to Wasm.
import 'dart:ffi' as ffi;

Error: Failed to compile application for the Web.
```

### Code sample

No code preview. Just the terminal command:
`flutter clean; flutter build web --wasm --release`

### Screenshots or Video

Not required.

### Logs

Here are the full error logs.

```console
Target dart2wasm failed: ProcessException: Process exited abnormally with exit code 64:
*NOTE*: Compilation to WasmGC is experimental.
The support may change, or be removed, with no advance notice.


../../../../.pub-cache/hosted/pub.dev/vector_graphics_compiler-1.1.11+1/lib/src/svg/_tessellator_ffi.dart:6:1: Error: 'dart:ffi' can't be imported when
compiling to Wasm.
import 'dart:ffi' as ffi;
^
../../../../.pub-cache/hosted/pub.dev/vector_graphics_compiler-1.1.11+1/lib/src/svg/_path_ops_ffi.dart:6:1: Error: 'dart:ffi' can't be imported when compiling
to Wasm.
import 'dart:ffi' as ffi;
^
  Command: /Users/adityam/Documents/Dev/flutter/bin/cache/dart-sdk/bin/dart compile wasm --packages=.dart_tool/package_config.json
  --extra-compiler-option=--dart-sdk=/Users/adityam/Documents/Dev/flutter/bin/cache/dart-sdk
  --extra-compiler-option=--platform=/Users/adityam/Documents/Dev/flutter/bin/cache/flutter_web_sdk/kernel/dart2wasm_platform.dill
  --extra-compiler-option=--delete-tostring-package-uri=dart:ui --extra-compiler-option=--delete-tostring-package-uri=package:flutter
  --extra-compiler-option=--import-shared-memory --extra-compiler-option=--shared-memory-max-pages=32768 -Ddart.vm.product=true
  -DFLUTTER_WEB_CANVASKIT_URL=https://www.gstatic.com/flutter-canvaskit/edd8546116457bdf1c5bdfb13ecb9463d2bb5ed4/ -DFLUTTER_WEB_AUTO_DETECT=false
  -DFLUTTER_WEB_USE_SKIA=false -DFLUTTER_WEB_USE_SKWASM=true --extra-compiler-option=--depfile=/Users/adityam/Documents/Mobile
  Apps/Memorio/site/.dart_tool/flutter_build/5389598c2a36d064db55a8edf57e320f/dart2wasm.d -O4 --no-name-section -o /Users/adityam/Documents/Mobile
  Apps/Memorio/site/.dart_tool/flutter_build/5389598c2a36d064db55a8edf57e320f/main.dart.wasm /Users/adityam/Documents/Mobile
  Apps/Memorio/site/.dart_tool/flutter_build/5389598c2a36d064db55a8edf57e320f/main.dart
#0      RunResult.throwException (package:flutter_tools/src/base/process.dart:125:5)
#1      _DefaultProcessUtils.run (package:flutter_tools/src/base/process.dart:321:19)
<asynchronous suspension>
#2      Dart2WasmTarget.build (package:flutter_tools/src/build_system/targets/web.dart:329:5)
<asynchronous suspension>
#3      _BuildInstance._invokeInternal (package:flutter_tools/src/build_system/build_system.dart:875:9)
<asynchronous suspension>
#4      Future.wait.<anonymous closure> (dart:async/future.dart:524:21)
<asynchronous suspension>
#5      _BuildInstance.invokeTarget (package:flutter_tools/src/build_system/build_system.dart:813:32)
<asynchronous suspension>
#6      Future.wait.<anonymous closure> (dart:async/future.dart:524:21)
<asynchronous suspension>
#7      _BuildInstance.invokeTarget (package:flutter_tools/src/build_system/build_system.dart:813:32)
<asynchronous suspension>
#8      FlutterBuildSystem.build (package:flutter_tools/src/build_system/build_system.dart:642:16)
<asynchronous suspension>
#9      WebBuilder.buildWeb (package:flutter_tools/src/web/compile.dart:92:34)
<asynchronous suspension>
#10     BuildWebCommand.runCommand (package:flutter_tools/src/commands/build_web.dart:230:5)
<asynchronous suspension>
#11     FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1394:27)
<asynchronous suspension>
#12     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
<asynchronous suspension>
#13     CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
#14     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:372:9)
<asynchronous suspension>
#15     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
<asynchronous suspension>
#16     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:308:5)
<asynchronous suspension>
#17     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:130:9)
<asynchronous suspension>
#18     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
<asynchronous suspension>
#19     main (package:flutter_tools/executable.dart:93:3)
<asynchronous suspension>

Compiling lib/main.dart for the Web...                             24.7s
Error: Failed to compile application for the Web.
```

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.22.2, on macOS 14.4.1 23E224 darwin-arm64, locale en-US)
    • Flutter version 3.22.2 on channel stable at /Users/adityam/Documents/Dev/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (4 days ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/adityam/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.90.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Brave Browser 125.1.66.118
    ! Error: Browsing on the local area network for Aditya Majethia’s iPhone. Ensure the device is unlocked and attached with a cable or associated with the
      same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by dcharkes** [comment]
Hi @adifyr!

`dart:ffi` support in WASM is highly experimental. (Only a limited subset of features is supported, and there are no guardrails! Moreover, without the multiple memory proposal, only one C library compiled to WASM can be used at once.) Implementing `dart:ffi` support in dart2wasm is tracked in:

* https://github.com/dart-lang/sdk/issues/46690

We're happy to receive contributions to dart2wasm which implement more features of `dart:ffi` and beef up the tests!

The error message you're seeing is from https://dart-review.googlesource.com/c/sdk/+/369020 (which was reverted). Once that gets relanded, we will land an experiment flag in Dart to allow experimentation https://dart-review.googlesource.com/c/sdk/+/370063). cc @eyebrowsoffire @mkustermann We might also need an experiment flag in Flutter.

**Comment by adifyr** [comment]
Hi @dcharkes. Thanks for the updated. I'm not using `dart:ffi` in any of my code. Maybe one of the dependencies is? Is it the `flutter_svg` package? How do I get my code to compile with WASM?

**Comment by mkustermann** [comment]
> Hi @dcharkes. Thanks for the update. I'm not using dart:ffi in any of my code. Maybe one of my dependencies has the import? Is it the flutter_svg package?
> 
> How do I get my code to compile with WASM?

@adifyr Ideally we make the conditional import be evaluated differently (which I'll see if that is possible). In the mean time I make a PR to `package:vector_math` that changes the condition, so it will evaluate the way we want.

**Comment by darshankawar** [comment]
I tried adding `flutter_svg` and ran the given command with latest stable and master using demo web app, but was unable to replicate the reported error, but based on the report, I'll label for tracking.

**Comment by Erluan** [comment]
I'm having the same issue with the 3.22.2. I was able to build in wasm in the 3.22.1 but when I updated, it was logging this error with the dart:ffi, and I dont import this lib at all.

**Comment by mkustermann** [comment]
I'm going to close this in favor of https://github.com/flutter/flutter/issues/149984

**Comment by adifyr** [comment]
Hi! Any updates on when the next hot-fix will be released to the flutter stable channel?

**Comment by mkustermann** [comment]
> Hi @mkustermann @dcharkes. Any ETA on when the next hot-fix will be released to the flutter stable channel?

@christopherfujino may know

**Comment by chandrabezzo** [comment]
@mkustermann I try using `3.23.0-13.0.pre.337`, but the issue still happen. I found the issue `Error: 'dart:ffi' can't be imported when compiling to Wasm.`. We need to improve the code for every package/plugin that we use?

**Comment by chandrabezzo** [comment]
But this issue fixed in 3.22.1

**Comment by mkustermann** [comment]
> @mkustermann I try using 3.23.0-13.0.pre.337, but the issue still happen. I found the issue Error: 'dart:ffi' can't be imported when compiling to Wasm.. We need to improve the code for every package/plugin that we use?

This error is correct, we do not want to allow importing `dart:ffi` at this moment, because we have very sketchy incomplete support for ffi in dart2wasm. It matters why you get the error, why is `dart:ffi` imported. (There was one issue where conditional imports on `if (dart.library.ffi)` were incorrectly evaluated to true, this has been fixed on main/master and will eventually make it to flutter stable).

> But this issue fixed in 3.22.1

flutter 3.22.1 is based on flutter/engine@55eae6864b296dd9f43b2cc7577ec256e5c32a8d & dart-lang/sdk@b0e7f171c7c9ab82944ffd6918b80a0df36e0549 (which is dart 3.4.1 I believe). This version doesn't have the bug fixes in (this version accidentally allowed importing `dart:ffi` which isn't actually working properly in dart2wasm).

Locking this issue.
